facilitator/src/transport/s3.rs

```diff
@@ -394,20 +456,41 @@ impl TransportWriter for MultipartUploadWriter {
 
         // Ignore output for now, but we might want the e_tag to check the
         // digest
-        self.runtime
-            .block_on(
-                self.client
-                    .complete_multipart_upload(CompleteMultipartUploadRequest {
-                        bucket: self.bucket.to_string(),
-                        key: self.key.to_string(),
-                        upload_id: self.upload_id.clone(),
-                        multipart_upload: Some(CompletedMultipartUpload {
-                            parts: Some(mem::take(&mut self.completed_parts)),
-                        }),
-                        ..Default::default()
+        let mut retries = 0;
+        let completed_parts = mem::take(&mut self.completed_parts);
+        loop {
```

Comment:
The duplication of these retry loops strongly suggests factoring out a "do_with_retries" function that takes a closure, and runs it in the loop specified below.

Reply:
This was less painful than I expected it to be and it's much better this way.